include/cuda/memory_resource

```diff
@@ -0,0 +1,529 @@
+  using _AsyncDeallocFn = void (*)(void*, void*, size_t, size_t, cuda::stream_ref);
+
+  _AsyncAllocFn __async_alloc_fn;
+  _AsyncDeallocFn __async_dealloc_fn;
+
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr _Async_alloc_vtable(_Alloc_vtable::_AllocFn __alloc_fn_,
+                                                              _Alloc_vtable::_DeallocFn __dealloc_fn_,
+                                                              _Alloc_vtable::_EqualFn __equal_fn_,
+                                                              _AsyncAllocFn __async_alloc_fn_,
+                                                              _AsyncDeallocFn __async_dealloc_fn_) noexcept
+      : _Alloc_vtable(__alloc_fn_, __dealloc_fn_, __equal_fn_)
+      , __async_alloc_fn(__async_alloc_fn_)
+      , __async_dealloc_fn(__async_dealloc_fn_)
+  {}
+};
+
+// clang-format off
+struct _Resource_vtable_builder
+{
+    template <class _Resource, class _Property>
+    _LIBCUDACXX_INLINE_VISIBILITY
+    static typename _Property::value_type _Get_property(void* __res) noexcept {
+        return get_property(*static_cast<const _Resource *>(__res), _Property{});
+    }
+
+    template <class _Resource>
+    _LIBCUDACXX_INLINE_VISIBILITY 
+    static void* _Alloc(void* __object, size_t __bytes, size_t __alignment) {
+        return static_cast<_Resource *>(__object)->allocate(__bytes, __alignment);
+    }
+
+    template <class _Resource>
+    _LIBCUDACXX_INLINE_VISIBILITY 
+    static void _Dealloc(void* __object, void* __ptr, size_t __bytes, size_t __alignment) {
+        return static_cast<_Resource *>(__object)->deallocate(__ptr, __bytes, __alignment);
+    }
+
+    template <class _Resource>
+    _LIBCUDACXX_INLINE_VISIBILITY 
+    static void* _Alloc_async(void* __object, size_t __bytes, size_t __alignment, cuda::stream_ref __stream) {
+        return static_cast<_Resource *>(__object)->allocate_async(__bytes, __alignment, __stream);
+    }
+
+    template <class _Resource>
+    _LIBCUDACXX_INLINE_VISIBILITY 
+    static void _Dealloc_async(void* __object, void* __ptr, size_t __bytes, size_t __alignment, cuda::stream_ref __stream) {
+        return static_cast<_Resource *>(__object)->deallocate_async(__ptr, __bytes, __alignment, __stream);
+    }
+
+    template <class _Resource>
+    _LIBCUDACXX_INLINE_VISIBILITY 
+    static bool _Equal(void* __left, void* __right) {
+        return *static_cast<_Resource *>(__left) == *static_cast<_Resource *>(__right);
+    }
+
+    _LIBCUDACXX_TEMPLATE(class _Resource, _AllocType _Alloc_type)
+      (requires(_Alloc_type == _AllocType::_Default)) //
+    _LIBCUDACXX_INLINE_VISIBILITY static constexpr _Async_alloc_vtable _Create() noexcept
+    {
+      return {&_Resource_vtable_builder::_Alloc<_Resource>,
+              &_Resource_vtable_builder::_Dealloc<_Resource>,
+              &_Resource_vtable_builder::_Equal<_Resource>,
+              nullptr,
+              nullptr};
+    }
```

Comment:
What do we need this for? Shouldn't we rather use the `_Alloc_vtable` then?

Reply:
You answered the two comments in a contradictory way.

Reply:
How so? I guess there was some misunderstanding.

Reply:
Yeah, I changed it to always return a `_Alloc_vtable`